Sources/SwiftProtobuf/ProtobufTextEncoding.swift

```diff
@@ -0,0 +1,291 @@
+        text.append(newText)
+    }
+    mutating func appendTokens(tokens: [ProtobufTextToken]) {
+        for t in tokens {
+            switch t {
+            case .beginArray: append(text: "[")
+            case .beginObject: append(text: "{")
+            case .boolean(let v):
+                // Note that quoted boolean map keys get stored as .string()
+                putBoolValue(value: v, quote: false)
+            case .colon: append(text: ":")
+            case .comma: append(text: ",")
+            case .endArray: append(text: "]")
+            case .endObject: append(text: "}")
+            case .null: putNullValue()
+            case .number(let v): append(text: v)
+            case .string(let v): putStringValue(value: v)
+            }
+        }
+    }
+    mutating func startField(name: String, tabLevel: Int, dropColon:Bool = false) {
+        for _ in 0..<tabLevel {
+            append(text:"  ")
+        }
+        
+        if dropColon {
+            append(text: name + " ")
+        } else {
+            append(text: name + ": ")
+        }
+    }
+    mutating func endField() {
+        append(text: "\n")
+    }
+    public mutating func startObject() {
+        append(text: "{\n")
+    }
+    public mutating func endObject(tabLevel: Int) {
+        for _ in 0..<tabLevel {
+            append(text:"  ")
+        }
+
+        append(text: "}")
+    }
+    mutating func putNullValue() {
+        append(text: "null")
+    }
+    mutating func putFloatValue(value: Float, quote: Bool) {
+        putDoubleValue(value: Double(value), quote: quote)
+    }
+    mutating func putDoubleValue(value: Double, quote: Bool) {
+        if value.isNaN {
+            append(text: "\"NaN\"")
+        } else if !value.isFinite {
+            if value < 0 {
+                append(text: "\"-Infinity\"")
+            } else {
+                append(text: "\"Infinity\"")
+            }
+        } else {
+            // TODO: Be smarter here about choosing significant digits
+            // See: protoc source has C++ code for this with interesting ideas
+            let s: String
+            if value < Double(Int64.max) && value > Double(Int64.min) && value == Double(Int64(value)) {
+                s = String(Int64(value))
+            } else {
+                s = String(value)
+            }
+            if quote {
+                append(text: "\"" + s + "\"")
+            } else {
+                append(text: s)
+            }
+        }
+    }
+    mutating func putInt64(value: Int64, quote: Bool) {
+        append(text: String(value))
+    }
+    mutating func putUInt64(value: UInt64, quote: Bool) {
+        append(text: String(value))
+    }
+    
+    mutating func putBoolValue(value: Bool, quote: Bool) {
+        if quote {
+            append(text: value ? "\"true\"" : "\"false\"")
+        } else {
+            append(text: value ? "true" : "false")
+        }
+    }
+    mutating func putStringValue(value: String) {
```

Comment:
The C++ code calls this `CEscapeAndAppend` function, which prints a C-escaped UTF-8 version of the string: https://github.com/google/protobuf/blob/fd046f6263fb17383cafdbb25c361e3451c31105/src/google/protobuf/stubs/strutil.cc#L579